aspnetcore/security/authentication/configure-oidc-web-authentication.md

```diff
@@ -0,0 +1,319 @@
+
+```csharp
+builder.Services.AddAuthentication(options =>
+{
+	options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+	options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
+})
+.AddCookie()
+.AddOpenIdConnect(options =>
+{
+	configuration.GetSection("OpenIDConnectSettings").Bind(options);
+	
+	options.Authority = configuration["OpenIDConnectSettings:Authority"];
+	options.ClientId = configuration["OpenIDConnectSettings:ClientId"];
+	options.ClientSecret = configuration["OpenIDConnectSettings:ClientSecret"];
+
+	options.SignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+	options.ResponseType = OpenIdConnectResponseType.Code;
+
+	options.SaveTokens = true;
+	options.GetClaimsFromUserInfoEndpoint = true;
+	options.TokenValidationParameters = new TokenValidationParameters
+	{
+		NameClaimType = "name"
+	};
+});
+```
+
+See [Secure an ASP.NET Core Blazor Web App with OpenID Connect (OIDC)](xref:blazor/security/blazor-web-app-oidc) for details on the different OpenID Connect options.
+
+See [Mapping, customizing, and transforming claims in ASP.NET Core](xref:security/authentication/claims) for the different claims mapping possibilities.
+
+> [!NOTE]
+> The following namespaces are required:
+
+```csharp
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.AspNetCore.Authentication.OpenIdConnect;
+using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+using Microsoft.IdentityModel.Tokens;
+```
+
+### Setup the configuration properties
+
+Add the OpenID Connect client settings to the application configuration properties. The settings must match the client configuration in the OpenID Connect server. No secrets are persisted in the application settings, the secrets are stored in a Key Vault in production environments or in user secrets in a development environment.
+
+```json
+"OpenIDConnectSettings": {
+    // OpenID Connect URL
+	"Authority": "https://localhost:44318",
+	// client ID from the OpenID Connect server
+	"ClientId": "oidc-pkce-confidential",
+	//"ClientSecret": "--stored-in-user-secrets-or-key-vault--"
+},
+```
+
+### Update the ASP.NET Core pipeline method in the program class.
+
+The UseRouting must be implemented before the UseAuthorization method.
+
+```csharp
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+app.UseAuthorization();
+app.MapRazorPages();
+```
+
+### Force authorization
+
+Add the Authorize attribute to the protected razor pages, for example the Index.cshtml.cs file
+
+```csharp
+[Authorize]
+```
+
+A better way would be to force the whole application to be authorized and opt out for unsecure pages
+
+```csharp
+builder.services.AddRazorPages().AddMvcOptions(options =>
+{
+    var policy = new AuthorizationPolicyBuilder()
+        .RequireAuthenticatedUser()
+        .Build();
+    options.Filters.Add(new AuthorizeFilter(policy));
```

Comment:
I think it would probably be better to demonstrate setting a [FallbackPolicy](https://learn.microsoft.com/aspnet/core/security/authorization/secure-data?view=aspnetcore-8.0#require-authenticated-users] over MVC filters for this purpose so it would also work if people add SignalR, Blazor, minimal APIs, etc... to the project.

Reply:
fixed